keys/src/webauthn_public_key.rs

```diff
@@ -0,0 +1,160 @@
+use std::{cmp::Ordering, convert::TryInto, fmt, str::FromStr};
+
+use hex::FromHex;
+
+use crate::{
+    errors::{KeysError, ParseError},
+    Signature,
+};
+
+#[derive(Clone, Copy)]
+#[cfg_attr(feature = "serde-derive", derive(nimiq_hash_derive::SerializeContent))]
+pub struct WebauthnPublicKey(pub p256::EncodedPoint);
+
+impl WebauthnPublicKey {
+    pub const SIZE: usize = 33;
+
+    pub fn verify(&self, signature: &Signature, data: &[u8]) -> bool {
+        let signature = p256::ecdsa::Signature::from_slice(&signature.to_bytes()).unwrap();
```

Comment:
I think I prefer a more clear-cut separation between them, i.e., having a separate struct for each sig type.

Reply:
I implemented a separate `ES256Signature` in this PR's branch and did the dublication/separation wherever necessary. But I think it's not clean. It also leads to a problem later, when we want to use the `WebauthnSignatureProof` also for Ed25519 keys and signatures (because apparently that is in the spec and is supported by YubiKey Series 5 [I ordered one and will try it]). Then we need to add an additional ident byte in the signature proof that distinguishes the two key types, and also the two signature types.

I implemented another approach in this commit: https://github.com/nimiq/core-rs-albatross/commit/50abce07287f682e8f8720ce9c1ff7da530083d7
It generalizes the `Signature` to be a simple typed array, just like an `Address` is internally represented by. It then has getter methods to get it as a Ed25519 signature or a P256 signature.

What do you think about that solution? I think it works very well, and we can get rid of the duplicate implementation and matching between the two versions, leading to cleaner and easier understood code.

Reply:
I checked the commit, but I don't particularly like it either.
The reason is that semantically speaking, the two signatures are constructed entirely differently (on a mathematical level). Also, it is rather a coincidence that both have the same length.

My original idea was to have structs for each variant (`Ed25519PublicKey`, `Ed25519Signature`,...) and then enums unifying those (`PublicKey`, `Signature`,...) so that we can have type safety over which can interact with which.
Also, so far, we've usually been hiding the concrete implementation as much as possible, so that we can swap out libraries if necessary. Thus, methods like `pub fn to_ed25519(&self) -> ed25519_zebra::Signature` should be avoided.

Reply:
This is now implemented in a single `SignatureProof` that has support for any curve public key and signature, as well as optional webauthn fields. Through the use of a single "type and flags" byte at the start of the serialization, this is efficiently serialized with only one additional byte, compared to before.